src/asm.rs

```diff
@@ -1,254 +1,544 @@
+                        constraint
+                    } else {
+                        // left for the next pass
+                        continue
+                    };
+
+                    // Rustc frontend guarantees that input and output types are "compatible",
+                    // so we can just use input var's type for the output variable.
+                    //
+                    // This decision is also backed by the fact that LLVM needs in and out 
+                    // values to be of *exactly the same type*, not just "compatible". 
+                    // I'm not sure if GCC is so picky too, but better safe than sorry.
+                    let ty = in_value.layout.gcc_type(self.cx, false);
+                    let tmp_var = self.current_func().new_local(None, ty, "output_register");
+
+                    // If the out_place is None (i.e `inout(reg) var` syntax was used), we translate
+                    // it to one "readwrite (+) output variable", otherwise we translate it to two 
+                    // "out and tied in" vars as described above.
+                    if out_place.is_none() {
+                        outputs.push(AsmOutOperand {
+                            constraint, 
+                            rust_idx,
+                            late,
+                            readwrite: true,
+                            tmp_var, 
+                            out_place,
+                        });
+                    } else {
+                        outputs.push(AsmOutOperand {
+                            constraint, 
+                            rust_idx,
+                            late,
+                            readwrite: false,
+                            tmp_var, 
+                            out_place,
+                        });
+
+                        let out_gcc_idx = outputs.len() - 1;
+                        let constraint = Cow::Owned(out_gcc_idx.to_string());
+
+                        inputs.push(AsmInOperand {
+                            constraint, 
+                            rust_idx, 
+                            val: in_value.immediate()
+                        });
+                    }
+                }
+
+                InlineAsmOperandRef::Const { ref string } => {
+                    constants_len += string.len() + att_dialect as usize;
+                }
+
+                InlineAsmOperandRef::SymFn { instance } => {
+                    constants_len += self.tcx.symbol_name(instance).name.len();
+                }
+                InlineAsmOperandRef::SymStatic { def_id } => {
+                    constants_len += self.tcx.symbol_name(Instance::mono(self.tcx, def_id)).name.len();
                 }
-                _ => {}
             }
         }
 
-        // All output operands must come before the input operands, hence the 2 loops.
-        for (idx, op) in operands.iter().enumerate() {
+        for (rust_idx, op) in rust_operands.iter().enumerate() {
```

Comment:
Since this function is big, I'd see value in adding a comment for each step. Here, that would be:
```rust
// 2. Register variables.
```
before this for loop.

Please do the same for the other steps.